File: src/main/java/net/vortetty/pulaskisandshaxes/items/PulaskiItem.java
package net.vortetty.pulaskisandshaxes.items;

import net.minecraft.block.*;
import net.minecraft.entity.EquipmentSlot;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.item.MiningToolItem;
import net.minecraft.item.ToolMaterial;
import net.minecraft.sound.SoundCategory;
import net.minecraft.sound.SoundEvents;
import net.minecraft.util.ActionResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.world.World;
import net.vortetty.pulaskisandshaxes.mixin.AxeItemAccessor;
import net.vortetty.pulaskisandshaxes.mixin.HoeItemAccessor;

import java.util.Arrays;
import java.util.Map;

public class PulaskiItem extends MiningToolItem {
	public static final Map<Block, Block> STRIPPED_BLOCKS;
	public static final Map<Block, BlockState> TILLED_BLOCKS;
	
	ToolMaterial material;
	
	public PulaskiItem(ToolMaterial material, float attackDamage, float attackSpeed, Settings settings) {
		super(attackDamage, attackSpeed, material, AxeItemAccessor.getEffectiveBlocks(), settings);
		this.material = material;
	}
	
	public float getMiningSpeed(ItemStack stack, BlockState state) {
		Material material = state.getMaterial();
		return this.miningSpeed;
	}
	
	public ActionResult useOnBlock(ItemUsageContext context) {
		World world = context.getWorld();
		BlockPos blockPos = context.getBlockPos();
		BlockState blockState = world.getBlockState(blockPos);
		Block block = (Block)STRIPPED_BLOCKS.get(blockState.getBlock());
		if (block != null) {
			PlayerEntity playerEntity = context.getPlayer();
			world.playSound(playerEntity, blockPos, SoundEvents.ITEM_AXE_STRIP, SoundCategory.BLOCKS, 1.0F, 1.0F);
			//stack.getOrCreateTag().putFloat("HOEING", 0f);
			if (!world.isClient) {
				world.setBlockState(blockPos, (BlockState)block.getDefaultState().with(PillarBlock.AXIS, blockState.get(PillarBlock.AXIS)), 11);
				if (playerEntity != null) {
					context.getStack().damage(1, playerEntity, (p) -> {
						p.sendToolBreakStatus(context.getHand());
					});
				}
			}
			return ActionResult.SUCCESS;
			
		} else if (context.getSide() != Direction.DOWN && (world.getBlockState(blockPos.up()).getMaterial()==Material.AIR||world.getBlockState(blockPos.up()).getMaterial()==Material.BAMBOO||world.getBlockState(blockPos.up()).getMaterial()==Material.BAMBOO_SAPLING||world.getBlockState(blockPos.up()).getMaterial()==Material.PLANT)) {
			BlockState blockState1 = (BlockState)TILLED_BLOCKS.get(world.getBlockState(blockPos).getBlock());
			if (blockState1 != null) {
				PlayerEntity playerEntity = context.getPlayer();
				world.playSound(playerEntity, blockPos, SoundEvents.ITEM_HOE_TILL, SoundCategory.BLOCKS, 1.0F, 1.0F);
				//stack.getOrCreateTag().putFloat("HOEING", 1f);
				if (!world.isClient) {
					world.setBlockState(blockPos, blockState1, 11);
					if (playerEntity != null) {
						context.getStack().damage(1, playerEntity, (p) -> {
							p.sendToolBreakStatus(context.getHand());
						});
					}
				}
				
				return ActionResult.SUCCESS;
			}
		}
		return ActionResult.PASS;
	}
	
	public boolean postHit(ItemStack stack, LivingEntity target, LivingEntity attacker) {
		stack.damage(2, attacker, (e) -> {
			e.sendEquipmentBreakStatus(EquipmentSlot.MAINHAND);
		});
		return true;
	}
	
	public boolean postMine(ItemStack stack, World world, BlockState state, BlockPos pos, LivingEntity miner) {
		if (!world.isClient && state.getHardness(world, pos) != 0.0F) {
			stack.damage(1, miner, (e) -> {
				e.sendEquipmentBreakStatus(EquipmentSlot.MAINHAND);
			});
		}
		//stack.getOrCreateTag().putFloat("HOEING", 0f);
		return true;
	}
	
	static {
		AxeItemAccessor.getEffectiveBlocks().addAll(Arrays.asList(Blocks.STRIPPED_ACACIA_LOG,Blocks.STRIPPED_ACACIA_WOOD,Blocks.STRIPPED_BIRCH_WOOD,Blocks.STRIPPED_BIRCH_LOG,Blocks.STRIPPED_CRIMSON_HYPHAE,Blocks.STRIPPED_CRIMSON_STEM,Blocks.STRIPPED_DARK_OAK_LOG,Blocks.STRIPPED_DARK_OAK_WOOD,Blocks.STRIPPED_JUNGLE_LOG,Blocks.STRIPPED_JUNGLE_WOOD,Blocks.STRIPPED_OAK_LOG,Blocks.STRIPPED_OAK_WOOD,Blocks.STRIPPED_SPRUCE_LOG,Blocks.STRIPPED_SPRUCE_WOOD,Blocks.STRIPPED_WARPED_HYPHAE,Blocks.STRIPPED_WARPED_STEM));
		STRIPPED_BLOCKS = AxeItemAccessor.getStrippedBlocks();
		TILLED_BLOCKS = HoeItemAccessor.getTilledBlocks();
	}
}

File: src/main/kotlin/net/vortetty/pulaskisandshaxes/items/PulaskiItem.kt
package net.vortetty.pulaskisandshaxes.items

import com.google.common.collect.BiMap
import com.mojang.datafixers.util.Pair
import net.minecraft.advancement.criterion.Criteria
import net.minecraft.block.*
import net.minecraft.entity.LivingEntity
import net.minecraft.item.*
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvents
import net.minecraft.tag.BlockTags
import net.minecraft.util.ActionResult
import net.minecraft.util.math.Direction
import net.minecraft.world.WorldEvents
import net.vortetty.pulaskisandshaxes.mixin.AxeItemAccessor
import net.vortetty.pulaskisandshaxes.mixin.HoeItemAccessor
import java.util.*
import java.util.function.Consumer
import java.util.function.Predicate

open class PulaskiItem(material: ToolMaterial?, attackDamage: Float, attackSpeed: Float, settings: Settings?) : MiningToolItem(attackDamage, attackSpeed, material, BlockTags.AXE_MINEABLE, settings) {
    private var STRIPPED_BLOCKS: Map<Block, Block> = AxeItemAccessor.getStrippedBlocks()
    private var TILLED_BLOCKS: Map<Block, Pair<Predicate<ItemUsageContext>, Consumer<ItemUsageContext>>> = HoeItemAccessor.getTilledBlocks()

    override fun useOnBlock(context: ItemUsageContext): ActionResult {
        val world = context.world
        val blockPos = context.blockPos
        val playerEntity = context.player
        val blockState = world.getBlockState(blockPos)
        val optional = getStrippedState(blockState)
        val optional2 = Oxidizable.getDecreasedOxidationState(blockState)
        val optional3 = Optional.ofNullable((HoneycombItem.WAXED_TO_UNWAXED_BLOCKS.get() as BiMap<*, *>)[blockState.block] as Block?).map { block: Block -> block.getStateWithProperties(blockState) }
        val itemStack = context.stack
        var optional4 = Optional.empty<BlockState>()
        val pair: Pair<Predicate<ItemUsageContext>, Consumer<ItemUsageContext>>? = TILLED_BLOCKS[world.getBlockState(blockPos).block]
        if (optional.isPresent) {
            world.playSound(playerEntity, blockPos, SoundEvents.ITEM_AXE_STRIP, SoundCategory.BLOCKS, 1.0f, 1.0f)
            optional4 = optional
        } else if (optional2.isPresent) {
            world.playSound(playerEntity, blockPos, SoundEvents.ITEM_AXE_SCRAPE, SoundCategory.BLOCKS, 1.0f, 1.0f)
            world.syncWorldEvent(playerEntity, WorldEvents.BLOCK_SCRAPED, blockPos, 0)
            optional4 = optional2
        } else if (optional3.isPresent) {
            world.playSound(playerEntity, blockPos, SoundEvents.ITEM_AXE_WAX_OFF, SoundCategory.BLOCKS, 1.0f, 1.0f)
            world.syncWorldEvent(playerEntity, WorldEvents.WAX_REMOVED, blockPos, 0)
            optional4 = optional3
        }
        return if (optional4.isPresent) {
            if (playerEntity is ServerPlayerEntity) {
                Criteria.ITEM_USED_ON_BLOCK.trigger(playerEntity as ServerPlayerEntity?, blockPos, itemStack)
            }
            world.setBlockState(blockPos, optional4.get(), Block.NOTIFY_ALL or Block.REDRAW_ON_MAIN_THREAD)
            if (playerEntity != null) {
                itemStack.damage(1, playerEntity as LivingEntity) { p: LivingEntity -> p.sendToolBreakStatus(context.hand) }
            }
            ActionResult.success(world.isClient)
        } else {
            return if (pair == null) {
                ActionResult.PASS
            } else {
                val predicate: Predicate<ItemUsageContext> = pair.first as Predicate<ItemUsageContext>
                val consumer: Consumer<ItemUsageContext> = pair.second as Consumer<ItemUsageContext>
                if (predicate.test(context)) {
                    world.playSound(playerEntity, blockPos, SoundEvents.ITEM_HOE_TILL, SoundCategory.BLOCKS, 1.0f, 1.0f)
                    if (!world.isClient) {
                        consumer.accept(context)
                        if (playerEntity != null) {
                            context.stack.damage(1, playerEntity as LivingEntity) { p: LivingEntity -> p.sendToolBreakStatus(context.hand) }
                        }
                    }
                    ActionResult.success(world.isClient)
                } else {
                    ActionResult.PASS
                }
            }
        }
    }

    private fun getStrippedState(state: BlockState): Optional<BlockState> {
        return Optional.ofNullable(STRIPPED_BLOCKS[state.block]).map { block: Block -> block.defaultState.with(PillarBlock.AXIS, state.get(PillarBlock.AXIS) as Direction.Axis) as BlockState }
    }

}
